python/ray/data/_internal/stats.py

```diff
@@ -310,94 +328,217 @@ def _summarize_iter(self) -> str:
+        if memory_stats:
             out += indent
-            memory_stats = [
-                round(e.max_rss_bytes / (1024 * 1024), 2) for e in exec_stats
-            ]
             out += "* Peak heap memory usage (MiB): {} min, {} max, {} mean\n".format(
-                min(memory_stats),
-                max(memory_stats),
-                int(np.mean(memory_stats)),
+                memory_stats["min"],
+                memory_stats["max"],
+                memory_stats["mean"],
             )
 
-        output_num_rows = [m.num_rows for m in blocks if m.num_rows is not None]
-        if output_num_rows:
+        output_num_rows_stats = stage_stats.get(DatasetStats.OUTPUT_NUM_ROWS, {})
+        if output_num_rows_stats:
             out += indent
             out += "* Output num rows: {} min, {} max, {} mean, {} total\n".format(
-                min(output_num_rows),
-                max(output_num_rows),
-                int(np.mean(output_num_rows)),
-                sum(output_num_rows),
+                output_num_rows_stats["min"],
+                output_num_rows_stats["max"],
+                output_num_rows_stats["mean"],
+                output_num_rows_stats["sum"],
             )
 
-        output_size_bytes = [m.size_bytes for m in blocks if m.size_bytes is not None]
-        if output_size_bytes:
+        output_size_bytes_stats = stage_stats.get(DatasetStats.OUTPUT_SIZE_BYTES, {})
+        if output_size_bytes_stats:
             out += indent
             out += "* Output size bytes: {} min, {} max, {} mean, {} total\n".format(
-                min(output_size_bytes),
-                max(output_size_bytes),
-                int(np.mean(output_size_bytes)),
-                sum(output_size_bytes),
+                output_size_bytes_stats["min"],
+                output_size_bytes_stats["max"],
+                output_size_bytes_stats["mean"],
+                output_size_bytes_stats["sum"],
             )
 
-        if exec_stats:
-            node_counts = collections.defaultdict(int)
-            for s in exec_stats:
-                node_counts[s.node_id] += 1
+        node_count_stats = stage_stats.get(DatasetStats.NODE_COUNT, {})
+        if node_count_stats:
             out += indent
             out += "* Tasks per node: {} min, {} max, {} mean; {} nodes used\n".format(
-                min(node_counts.values()),
-                max(node_counts.values()),
-                int(np.mean(list(node_counts.values()))),
-                len(node_counts),
+                node_count_stats["min"],
+                node_count_stats["max"],
+                node_count_stats["mean"],
+                node_count_stats["count"],
             )
-
         return out
 
+    def get_total_wall_time(self) -> float:
+        """Returns the sum of maximum wall time per stage over all stages."""
+        total_wall_time = 0
+        for stage_name, ss in self.stage_stats.items():
```

Comment:
This is implicitly expecting that the stage stats has already been calculated, which is a bit of a concerning assumption for the API. Similar to what I commented above, why not have this function take in a stage stats dict (or a list thereof) to make this dependency explicit?

Reply:
@clarkzinzow - We need to have `get_total_wall_time()` and `get_total_cpu_time()`, `get_max_heap_memory()` for all stages. An example usage:

```
ds = ray.data.read_images(...)
ds = ds.map_batches(...)
stats = ds._get_stats_object()
total_cpu_time = stats.get_total_cpu_time()
```

If passing a list of stage stats dict, it will make the API hard to use, for getting aggregated stats across stages. Can we still keep the API simple?

Reply:
@c21 Agreed that the UX here needs to be better, although I think that your previous solution would have all returned 0, since the calculation of the stage stats isn't triggered in that example usage.

I'm wondering if we're encountering a lot of pain here because we're trying to make a complex internal class, `DatasetStats`, a user-facing/dev-facing abstraction without simplifying it; it has a data model and API geared towards the collection of the stats, in addition to printing them. What we basically want is a user-facing/dev-facing struct-like object that hides a lot of the complexity/internal details of `DatasetStats` and that has some helper APIs (e.g. get total CPU time across all stages) for introspecting and printing the stats. Why don't we introduce some new dataclasses that serve as stats containers, with APIs for introspecting and summarizing the stats?

I think that a class hierarchy similar to the following could work:

```python
@dataclass
class DatasetStatsSummary:
    stages_stats: List[StageStatsSummary]
    iter_stats: IterStatsSummary
    parents: List[DatasetStatsSummary]

    def to_string(self, already_printed: Optional[Set[str]] = None, include_parent: bool = True) -> str:
        # Current DatasetStats.summary_string() implementation should go here, sans the read stats fetching.

    def get_total_wall_time(self) -> float:
        return sum(
            stage_stats.wall_time.get("max", 0)
            for stage_stats in self.stages_stats
        )

    def get_total_cpu_time(self) -> float:
        return sum(
            stage_stats.cpu_time.get("sum", 0)
            for stage_stats in self.stages_stats
        )

    def get_max_heap_memory(self) -> float:
        return max(
            stage_stats.peak_heap_memory.get("max", 0)
            for stage_stats in self.stages_stats
        )


@dataclass
class StageStatsSummary:
    block_execution_summary: str
    wall_time: Dict[str, float]
    cpu_time: Dict[str, float]
    peak_heap_memory: Dict[str, float]
    output_num_rows: Dict[str, float]
    output_size_bytes: Dict[str, float]
    node_count: Dict[str, float]
    is_substage: bool

    @classmethod
    def from_block_metadata(cls, block_metas: List[BlockMetadata], is_substage: bool) -> StageStatsSummary:
        # Current DatasetStats._calculate_block_stats() implementation should go here.

    def __str__(self) -> str:
        # Current DatasetStats._summarize_blocks() implementation should go here.


@dataclass
class IterStatsSummary:
    wait_time: float
    get_time: float
    next_time: float
    format_time: float
    user_time: float
    total_time: float

    def __str__(self) -> str:
        # Current DatasetStats._summarize_iter() implementation should go here.


class DatasetStats:
    ...

    def to_summary(self) -> DatasetStatsSummary:
        if self.needs_stats_actor:
            # Get read block metadata from stats actor, put in `self.stages["read"]`.

        stages_stats = []
        for n, (stage_name, metadata) in enumerate(self.stages.items()):
            stages_stats.append(StageStatsSummary.from_block_metadata(metadata, is_substage=False))
        iter_stats = IterStatsSummary(self.iter_wait_s.get(), ...)
        stats_summary_parents = [p.to_summary() for p in self.parents]
        return DatasetStatsSummary(stages_stats, iter_stats, stats_summary_parents)


# plan.py

class ExecutionPlan:
    def stats_summary(self) -> DatasetStatsSummary:
        self.execute()
        return self._snapshot_stats.to_summary()


# dataset.py

class Dataset:
    def _get_stats_summary(self) -> DatasetStatsSummary:
        return self._plan.stats_summary()

    def stats(self) -> str:
        return self._get_stats_summary().to_string()
```

What do y'all think?

Reply:
I think the overall refactoring/class structure with `Dataset/Stage/IterStatsSummary` makes sense. For methods like `DatasetStatsSummary.get_total_wall_time()`, I think we'd also need to access the parent's StatsSummary and include them in the max/sum calculations?

Reply:
The above `Dataset/Stage/IterStatsSummary` structure sounds good to me, which is cleaner than current solution.

How about starting with `DeveloperAPI` instead of `PublicAPI` for `DatasetStatsSummary`? So we have more room to tweak in next few releases.

Reply:
Agreed with starting with `DeveloperAPI`!

Reply:
@jianoaix - WDYT of above proposal?

Reply:
Sg to start with DeveloperAPI.